crucible/src/Lang/Crucible/Backend/Prove.hs

```diff
@@ -0,0 +1,135 @@
+{-|
+Module      : Lang.Crucible.Backend.Prove
+Description : Proving goals under assumptions
+Copyright   : (c) Galois, Inc 2024
+License     : BSD3
+-}
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Lang.Crucible.Backend.Prove
+  ( ProofResult(..)
+  , proveGoal
+  , proveProofGoal
+  , proveGoals
+  , proveObligations
+  , proveCurrentObligations
+  ) where
+
+import           Control.Lens ((^.))
+import           Data.Functor.Const (Const(Const, getConst))
+
+import qualified What4.Interface as W4
+import qualified What4.Expr as WE
+import qualified What4.Solver.Adapter as WSA
+import qualified What4.SatResult as W4R
+
+import qualified Lang.Crucible.Backend as CB
+import           Lang.Crucible.Backend.Assumptions (Assumptions)
+import           Lang.Crucible.Backend.ProofGoals (traverseGoalsWithAssumptions)
+
+type Goal sym = CB.ProofGoal (CB.Assumptions sym) (CB.Assertion sym)
+
+-- | The result of attempting to prove a goal with an SMT solver.
+--
+-- The constructors of this type correspond to those of 'W4R.SatResult'.
+data ProofResult sym t
+   = -- | The goal was proved
+     --
+     -- Corresponds to 'W4R.Unsat'.
+     Proved (Goal sym)
+     -- | The goal was disproved, and a model that falsifies it is available.
+     --
+     -- Corresponds to 'W4R.Sat'.
+   | Disproved (Goal sym) (WE.GroundEvalFn t) (Maybe (WE.ExprRangeBindings t))
+     -- | The SMT solver returned \"unknown\".
+     --
+     -- Corresponds to 'W4R.Unknown'.
+   | Unknown (Goal sym)
+
+-- | Prove a single goal.
```

Comment:
Perhaps say:

```suggestion
-- | Prove a single goal (i.e., an 'CB.Assertion') under the supplied 'Assumptions'.
```

It took me a bit to figure out what the difference between `proveGoal` and `proveProofGoal` is: it's the data types involved. Calling out the data types in the Haddocks makes this difference a little more obvious, in my opinion.